wifi: TWT improvements

Rex-Chen-NXP opened the pull request:
A. Add two new parameters for TWT setup
B. Fix  TWT quick setup: Add default exponent value
C. Add Broadcast TWT feature support for AP
D. Convert TWT command to use `getopt`

Rex-Chen-NXP committed:
net: wifi: shell: enhance twt parse parameters code

Enhance the parse parameters code for twt.

Signed-off-by: Rex Chen <rex.chen_1@nxp.com>

Rex-Chen-NXP committed:
net: wifi: shell: add btwt feature support

Add btwt_setup cmd for sap.

Signed-off-by: Rex Chen <rex.chen_1@nxp.com>

krish2718 commented:
Please update the title and add description to explain these set of changes.

Rex-Chen-NXP commented:
> Please update the title and add description to explain these set of changes.

Yes, of course, updated title and changes details , changes info also in separated PRs commit mesg

Rex-Chen-NXP commented:
Hi @krish2718 @jukkar 
Could you help to review?

jukkar commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -3184,15 +3200,16 @@ SHELL_STATIC_SUBCMD_SET_CREATE(wifi_twt_ops,
 		"<dialog_token: 1-255> <flow_id: 0-7> <responder: 0/1> <trigger: 0/1> <implicit:0/1> "
 		"<announce: 0/1> <twt_wake_interval: 1-262144us> <twt_interval: 1us-2^31us>.\n"
-		"<twt_wake_ahead_duration>: 0us-2^31us>\n",
+		"<twt_wake_ahead_duration>: 0us-2^31us>\n"
```
We are getting now quite many positional parameters for these commands and it looks like not very user friendly way to implement this. Could this changed to use parameter "keywords", so something like `-s, --setup <cmd> ....` so like what we have in other commands?

Rex-Chen-NXP replied:
Yes, of course,  will implement as comment.

Rex-Chen-NXP replied:
Hi @jukkar  Updated as comment, could you review again?

Rex-Chen-NXP commented:
Hi @krish2718  @jukkar 
Could you help review.
Thanks!

jukkar commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -1526,74 +1527,154 @@ static int cmd_wifi_twt_setup_quick(const struct shell *sh, size_t argc,
-	params.negotiation_type = (enum wifi_twt_negotiation_type)value;
+		case 'c':
+			if (!parse_number(sh, &value, state->optarg, NULL,
+			    WIFI_TWT_SETUP_CMD_REQUEST, WIFI_TWT_SETUP_CMD_DEMAND)) {
```
Minor nit, but the indentation is wrong here and also for the other added lines.
So here the code should look like
```
			if (!parse_number(sh, &value, state->optarg, NULL,
					  WIFI_TWT_SETUP_CMD_REQUEST,
					  WIFI_TWT_SETUP_CMD_DEMAND)) {
```
Is it possible to adjust your editor to indent the lines automatically?

Rex-Chen-NXP replied:
Thanks, all updated as comment

jukkar commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -1526,74 +1527,154 @@ static int cmd_wifi_twt_setup_quick(const struct shell *sh, size_t argc,
-	params.dialog_token = (uint8_t)value;
+		case 'f':
+			if (!parse_number(sh, &value, state->optarg, NULL, 0,
+			    (WIFI_MAX_TWT_FLOWS - 1))) {
```
indentation

Rex-Chen-NXP replied:
updated as comment

jukkar commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -1526,74 +1527,154 @@ static int cmd_wifi_twt_setup_quick(const struct shell *sh, size_t argc,
-	params.setup.announce = (bool)value;
+		case 'w':
+			if (!parse_number(sh, &value, state->optarg, NULL, 1,
+			    WIFI_MAX_TWT_WAKE_INTERVAL_US)) {
```
indentation

Rex-Chen-NXP replied:
updated as comment

jukkar commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -1526,74 +1527,154 @@ static int cmd_wifi_twt_setup_quick(const struct shell *sh, size_t argc,
-	params.setup.twt_wake_interval = (uint32_t)value;
+		case 'i':
+			if (!parse_number(sh, &value, state->optarg, NULL, 1,
+			    WIFI_MAX_TWT_INTERVAL_US)) {
```
indentation

Rex-Chen-NXP replied:
updated as comment

jukkar commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -1526,74 +1527,154 @@ static int cmd_wifi_twt_setup_quick(const struct shell *sh, size_t argc,
-		return -EINVAL;
+		case 'D':
+			if (!parse_number(sh, &value, state->optarg, NULL, 0,
+			  WIFI_MAX_TWT_WAKE_AHEAD_DURATION_US)) {
```
indentation

Rex-Chen-NXP replied:
updated as comment

jukkar commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -1526,74 +1527,154 @@ static int cmd_wifi_twt_setup_quick(const struct shell *sh, size_t argc,
+
+		case 'e':
+			if (!parse_number(sh, &value, state->optarg, NULL, 0,
+			    WIFI_MAX_TWT_EXPONENT)) {
```
indentation

Rex-Chen-NXP replied:
updated as comment

Rex-Chen-NXP commented:
Hi @jukkar @krish2718 
All updated as comments, pls help review.

krish2718 commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -1544,74 +1545,156 @@ static int cmd_wifi_twt_setup_quick(const struct shell *sh, size_t argc,
 
-static int cmd_wifi_twt_setup(const struct shell *sh, size_t argc,
-			      char *argv[])
+static int twt_args_to_params(const struct shell *sh, size_t argc, char *argv[],
```
can you please do the getopt conversion in a separate commit

Rex-Chen-NXP replied:
There is no plan to make the getopt conversion in a separate commit.
Actually one of the sub-commit almost all related with this twt_args_to_params, do you mean to make the fix of twt quick setup and twt tear down in  separate commit?

krish2718 replied:
basically 1 commit for getopt conversion and another commit for extending TWT params

krish2718 replied:
> Add parameters for twt setup and enhance the parse parameters code

`and` implies that these can split in two commits, can you please do that.

Rex-Chen-NXP replied:
There will be not clearly if separate getopt conversion  into a commit.
So I make the PR in 4 separated commits base on the commits function.

Rex-Chen-NXP replied:
Yes, of course, this is reasonable, separated as comment, also make quick setup and teardown in two separated commits to make code logic more clearly

krish2718 commented on drivers/wifi/nxp/nxp_wifi_drv.c:
```diff
@@ -1356,10 +1356,12 @@ static int nxp_wifi_set_twt(const struct device *dev, struct wifi_twt_params *pa
 		twt_setup_conf.implicit = params->setup.implicit;
 		twt_setup_conf.announced = params->setup.announce;
 		twt_setup_conf.trigger_enabled = params->setup.trigger;
+		twt_setup_conf.twt_info_disabled = params->setup.twt_info_disable;
```
commit log says quick setup but these are added to the full TWT command

Rex-Chen-NXP replied:
Have separated the quick setup and I-TWT as independence PRs

krish2718 commented on drivers/wifi/nxp/nxp_wifi_drv.c:
```diff
@@ -1374,6 +1374,25 @@ static int nxp_wifi_set_twt(const struct device *dev, struct wifi_twt_params *pa
+	btwt_config.twt_exponent = params->btwt.twt_exponent;
+	btwt_config.sp_gap = params->btwt.sp_gap;
+
+	ret = wlan_set_btwt_cfg(&btwt_config);
+
+	return ret;
```
`return wlan_set_btwt_cfg(&btwt_config);` unless you want to log the return value

Rex-Chen-NXP replied:
Updated as comment

krish2718 commented on drivers/wifi/nxp/nxp_wifi_drv.c:
```diff
@@ -1374,6 +1374,25 @@ static int nxp_wifi_set_twt(const struct device *dev, struct wifi_twt_params *pa
+static int nxp_wifi_set_btwt(const struct device *dev, struct wifi_twt_params *params)
+{
+	wlan_btwt_config_t btwt_config;
+	int ret = -1;
```
no need to initialize it would be overridden anwyay.

Rex-Chen-NXP replied:
Updated as comment

krish2718 commented on modules/hostap/src/supp_api.c:
```diff
@@ -1801,6 +1801,18 @@ int supplicant_set_twt(const struct device *dev, struct wifi_twt_params *params)
+	const struct wifi_mgmt_ops *const wifi_mgmt_api = get_wifi_mgmt_api(dev);
+
+	if (!wifi_mgmt_api || !wifi_mgmt_api->set_btwt) {
+		wpa_printf(MSG_ERROR, "Set TWT not supported");
```
`Set Broadcast TWT not supported`

Rex-Chen-NXP replied:
Updated as comment

krish2718 commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -1545,6 +1545,45 @@ static int cmd_wifi_twt_setup_quick(const struct shell *sh, size_t argc,
+
+	context.sh = sh;
+
+	params.btwt.sub_id = (uint16_t)strtol(argv[idx++], NULL, 10);
```
no error checks for parsing `strtol`?

Rex-Chen-NXP replied:
Updated as comment

Rex-Chen-NXP committed:
net: wifi: shell: add parameters for twt setup

Add two parameters for twt setup.

Signed-off-by: Rex Chen <rex.chen_1@nxp.com>

Rex-Chen-NXP commented:
Hi @krish2718  @jukkar  @MaochenWang1 
Separated PRs and updated as comments, pls help review.

MaochenWang1 commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -1544,74 +1544,139 @@ static int cmd_wifi_twt_setup_quick(const struct shell *sh, size_t argc,
+static int cmd_wifi_twt_setup(const struct shell *sh, size_t argc,
+			      char *argv[])
+{
+	struct net_if *iface = net_if_get_first_wifi();
```
Please use `net_if_get_wifi_sta();`

Rex-Chen-NXP replied:
Updated as comments.

Rex-Chen-NXP commented:
Hi @krish2718  @jukkar 
I want to insert other discussion about parameters usage of twt setup cmd.
The below two usage items, from the value range and spec description  , 'twt_wake_interval' should named as 'TWT Wake Duration', and 'twt_interval' should calculated from TWT Wake Interval(TWT Wake Interval Mantissa) and Exponent, so name as 'TWT Wake Interval' will be suitable. 
A. <twt_wake_interval: 1-262144us> 
B. <twt_interval: 1us-2^31us>

As I didn't find description about 'twt_interval' from spec, if my above suggestion and understanding is error, pls tell me.
If you agree, I will raise other PR to change the usage after this PR merged.
Thanks.

krish2718 commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -752,6 +752,10 @@ struct wifi_twt_params {
 			 * prepare the data before TWT SP starts.
 			 */
 			uint32_t twt_wake_ahead_duration;
+			/** TWT info enabled or disable */
```
can you please elaborate? there is a separate `TWT Information frame` action frame.

Rex-Chen-NXP replied:
Yes, it is.

krish2718 commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -752,6 +752,10 @@ struct wifi_twt_params {
+			/** TWT info enabled or disable */
+			bool twt_info_disable;
+			/** TWT exponent */
+			uint8_t exponent;
```
As the (mantissa, exponent) are protocol fields, it was intentionally decided to use only the `internal` in `ms` as a user input to keep things simple and in the driver we do https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/wifi/nrf_wifi/src/wifi_mgmt.c#L250 with exponent=0. I understand that exposing both (mantissa, exponent) directly gives us flexibility but would put burden on users as they are only concerned about times.

Rex-Chen-NXP replied:
Under the condition exponent=0, mantissa is equal to interval, but for 'TWT Parameter Set field format' descripted in spc, mantissa use 2 Octets,  not enough to represent twt_interval, still need to set exponent value, then there will be multiple sets of exponent and mantissa, and maybe twt_interval not equal to mantissa * 2^(exponent ) accurately

Rex-Chen-NXP replied:
So, the last discuss point for this PR is need add exponent parameter or not.
As twt feature support by vendor's different implementation. Could we add exponent parameter, if user input exponent value from shell, then vendor use the input value, or vendor use the default value 0?

krish2718 replied:
As I said for testing all possible combinations we need both as inputs, but I am thinking about whether user's are knowledgeable enough to understand how (mantissa, exponent) translate to the wake/sleep timings. Let's see what others think.

krish2718 commented:
> Hi @krish2718 @jukkar I want to insert other discussion about parameters usage of twt setup cmd. The below two usage items, from the value range and spec description , 'twt_wake_interval' should named as 'TWT Wake Duration', and 'twt_interval' should calculated from TWT Wake Interval(TWT Wake Interval Mantissa) and Exponent, so name as 'TWT Wake Interval' will be suitable. A. <twt_wake_interval: 1-262144us> B. <twt_interval: 1us-2^31us>
> 
> As I didn't find description about 'twt_interval' from spec, if my above suggestion and understanding is error, pls tell me. If you agree, I will raise other PR to change the usage after this PR merged. Thanks.

Please see https://github.com/zephyrproject-rtos/zephyr/pull/82312/files#r1875537287 , it's about keeping UI simple.

Rex-Chen-NXP commented:
> > Hi @krish2718 @jukkar I want to insert other discussion about parameters usage of twt setup cmd. The below two usage items, from the value range and spec description , 'twt_wake_interval' should named as 'TWT Wake Duration', and 'twt_interval' should calculated from TWT Wake Interval(TWT Wake Interval Mantissa) and Exponent, so name as 'TWT Wake Interval' will be suitable. A. <twt_wake_interval: 1-262144us> B. <twt_interval: 1us-2^31us>
> > As I didn't find description about 'twt_interval' from spec, if my above suggestion and understanding is error, pls tell me. If you agree, I will raise other PR to change the usage after this PR merged. Thanks.
> 
> Please see https://github.com/zephyrproject-rtos/zephyr/pull/82312/files#r1875537287 , it's about keeping UI simple.

Sure, even if we don't use exponent, but the range of 'twt_wake_interval' is matched with 'TWT Wake Duration', so should we change the usage to 'TWT Wake Duration', twt_wake_interval make user confuse.

krish2718 commented:
> > > Hi @krish2718 @jukkar I want to insert other discussion about parameters usage of twt setup cmd. The below two usage items, from the value range and spec description , 'twt_wake_interval' should named as 'TWT Wake Duration', and 'twt_interval' should calculated from TWT Wake Interval(TWT Wake Interval Mantissa) and Exponent, so name as 'TWT Wake Interval' will be suitable. A. <twt_wake_interval: 1-262144us> B. <twt_interval: 1us-2^31us>
> > > As I didn't find description about 'twt_interval' from spec, if my above suggestion and understanding is error, pls tell me. If you agree, I will raise other PR to change the usage after this PR merged. Thanks.
> > 
> > 
> > Please see https://github.com/zephyrproject-rtos/zephyr/pull/82312/files#r1875537287 , it's about keeping UI simple.
> 
> Sure, even if we don't use exponent, but the range of 'twt_wake_interval' is matched with 'TWT Wake Duration', so should we change the usage to 'TWT Wake Duration', twt_wake_interval make user confuse.

 `twt_wake_interval` is mapped to `Nominal Minimum TWT Wake Duration field` and in 802.11ax-2021, the max value is https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/net/wifi_mgmt.h#L765 i.e., `Wake Duration Unit subfield == 1` and `Nominal Minimum TWT Wake Duration field == 256` , can you please let me know how the range is incorrect?

Rex-Chen-NXP commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -3331,7 +3336,7 @@ SHELL_STATIC_SUBCMD_SET_CREATE(wifi_twt_ops,
 		"<negotiation_type, 0: Individual, 1: Broadcast, 2: Wake TBTT>\n"
 		"<setup_cmd: 0: Request, 1: Suggest, 2: Demand>\n"
-		"<dialog_token: 1-255> <flow_id: 0-7>.\n",
+		"<dialog_token: 1-255> <flow_id: 0-7> <teardown_all_twt: 0/1>.\n",
```
Okay, Will remove.

Rex-Chen-NXP replied:
Removed teardown changes.

Rex-Chen-NXP commented:
> > > > Hi @krish2718 @jukkar I want to insert other discussion about parameters usage of twt setup cmd. The below two usage items, from the value range and spec description , 'twt_wake_interval' should named as 'TWT Wake Duration', and 'twt_interval' should calculated from TWT Wake Interval(TWT Wake Interval Mantissa) and Exponent, so name as 'TWT Wake Interval' will be suitable. A. <twt_wake_interval: 1-262144us> B. <twt_interval: 1us-2^31us>
> > > > As I didn't find description about 'twt_interval' from spec, if my above suggestion and understanding is error, pls tell me. If you agree, I will raise other PR to change the usage after this PR merged. Thanks.
> > > 
> > > 
> > > Please see https://github.com/zephyrproject-rtos/zephyr/pull/82312/files#r1875537287 , it's about keeping UI simple.
> > 
> > 
> > Sure, even if we don't use exponent, but the range of 'twt_wake_interval' is matched with 'TWT Wake Duration', so should we change the usage to 'TWT Wake Duration', twt_wake_interval make user confuse.
> 
> `twt_wake_interval` is mapped to `Nominal Minimum TWT Wake Duration field` and in 802.11ax-2021, the max value is https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/net/wifi_mgmt.h#L765 i.e., `Wake Duration Unit subfield == 1` and `Nominal Minimum TWT Wake Duration field == 256` , can you please let me know how the range is incorrect? Please note that depending on the user value, the `Wake Duration Unit subfield == 1` is auto-encoded to 0/1 in the driver/firmware.

Yes, I mean should rename `twt_wake_interval` to

krish2718 commented:
> > > > > Hi @krish2718 @jukkar I want to insert other discussion about parameters usage of twt setup cmd. The below two usage items, from the value range and spec description , 'twt_wake_interval' should named as 'TWT Wake Duration', and 'twt_interval' should calculated from TWT Wake Interval(TWT Wake Interval Mantissa) and Exponent, so name as 'TWT Wake Interval' will be suitable. A. <twt_wake_interval: 1-262144us> B. <twt_interval: 1us-2^31us>
> > > > > As I didn't find description about 'twt_interval' from spec, if my above suggestion and understanding is error, pls tell me. If you agree, I will raise other PR to change the usage after this PR merged. Thanks.
> > > > 
> > > > 
> > > > Please see https://github.com/zephyrproject-rtos/zephyr/pull/82312/files#r1875537287 , it's about keeping UI simple.
> > > 
> > > 
> > > Sure, even if we don't use exponent, but the range of 'twt_wake_interval' is matched with 'TWT Wake Duration', so should we change the usage to 'TWT Wake Duration', twt_wake_interval make user confuse.
> > 
> > 
> > `twt_wake_interval` is mapped to `Nominal Minimum TWT Wake Duration field` and in 802.11ax-2021, the max value is https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/net/wifi_mgmt.h#L765 i.e., `Wake Duration Unit subfield == 1` and `Nominal Minimum TWT Wake Duration field == 256` , can you please let me know how the range is incorrect? Please note that depending on the user value, the `Wake Duration Unit subfield == 1` is auto-encoded to 0/1 in the driver/firmware.
> 
> Yes, I mean why don't named as 'TWT Wake Duration' directly?

Sure we can rename it to `twt_wake_duration`.

Rex-Chen-NXP commented:
> > > > > > Hi @krish2718 @jukkar I want to insert other discussion about parameters usage of twt setup cmd. The below two usage items, from the value range and spec description , 'twt_wake_interval' should named as 'TWT Wake Duration', and 'twt_interval' should calculated from TWT Wake Interval(TWT Wake Interval Mantissa) and Exponent, so name as 'TWT Wake Interval' will be suitable. A. <twt_wake_interval: 1-262144us> B. <twt_interval: 1us-2^31us>
> > > > > > As I didn't find description about 'twt_interval' from spec, if my above suggestion and understanding is error, pls tell me. If you agree, I will raise other PR to change the usage after this PR merged. Thanks.
> > > > > 
> > > > > 
> > > > > Please see https://github.com/zephyrproject-rtos/zephyr/pull/82312/files#r1875537287 , it's about keeping UI simple.
> > > > 
> > > > 
> > > > Sure, even if we don't use exponent, but the range of 'twt_wake_interval' is matched with 'TWT Wake Duration', so should we change the usage to 'TWT Wake Duration', twt_wake_interval make user confuse.
> > > 
> > > 
> > > `twt_wake_interval` is mapped to `Nominal Minimum TWT Wake Duration field` and in 802.11ax-2021, the max value is https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/net/wifi_mgmt.h#L765 i.e., `Wake Duration Unit subfield == 1` and `Nominal Minimum TWT Wake Duration field == 256` , can you please let me know how the range is incorrect? Please note that depending on the user value, the `Wake Duration Unit subfield == 1` is auto-encoded to 0/1 in the driver/firmware.
> > 
> > 
> > Yes, I mean why don't named as 'TWT Wake Duration' directly?
> 
> Sure we can rename it to `twt_wake_duration`. But it would be a breaking API change.

Okay, will raise PR later after this PR merged.

krish2718 commented:
> > > > > > > Hi @krish2718 @jukkar I want to insert other discussion about parameters usage of twt setup cmd. The below two usage items, from the value range and spec description , 'twt_wake_interval' should named as 'TWT Wake Duration', and 'twt_interval' should calculated from TWT Wake Interval(TWT Wake Interval Mantissa) and Exponent, so name as 'TWT Wake Interval' will be suitable. A. <twt_wake_interval: 1-262144us> B. <twt_interval: 1us-2^31us>
> > > > > > > As I didn't find description about 'twt_interval' from spec, if my above suggestion and understanding is error, pls tell me. If you agree, I will raise other PR to change the usage after this PR merged. Thanks.
> > > > > > 
> > > > > > 
> > > > > > Please see https://github.com/zephyrproject-rtos/zephyr/pull/82312/files#r1875537287 , it's about keeping UI simple.
> > > > > 
> > > > > 
> > > > > Sure, even if we don't use exponent, but the range of 'twt_wake_interval' is matched with 'TWT Wake Duration', so should we change the usage to 'TWT Wake Duration', twt_wake_interval make user confuse.
> > > > 
> > > > 
> > > > `twt_wake_interval` is mapped to `Nominal Minimum TWT Wake Duration field` and in 802.11ax-2021, the max value is https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/net/wifi_mgmt.h#L765 i.e., `Wake Duration Unit subfield == 1` and `Nominal Minimum TWT Wake Duration field == 256` , can you please let me know how the range is incorrect? Please note that depending on the user value, the `Wake Duration Unit subfield == 1` is auto-encoded to 0/1 in the driver/firmware.
> > > 
> > > 
> > > Yes, I mean why don't named as 'TWT Wake Duration' directly?
> > 
> > 
> > Sure we can rename it to `twt_wake_duration`. But it would be a breaking API change.
> 
> Okay, will raise PR later after this PR merged.

Please note we have to follow the rules here for any `stable` API https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html but I guess there aren't many users of TWT, so, not sure if we should call it `stable` or not. @jukkar WDYT?

Rex-Chen-NXP commented:
Hi @krish2718 @jukkar @MaochenWang1 
Follow @krish2718 comment to remove teardown changes, and there is no conclusion of 'exponent ', so keep exponent as shell input.

krish2718 commented:
> Hi @krish2718 @jukkar @MaochenWang1 
> Follow @krish2718 comment to remove teardown changes, and there is no conclusion of 'exponent ', so keep exponent as shell input.

Should we add both the options? If the user gives time we derive (exponent, mantissa) and if the user gives (exponent, mantissa) we use it directly, WDYT?

Rex-Chen-NXP commented:
> > Hi @krish2718 @jukkar @MaochenWang1
> > Follow @krish2718 comment to remove teardown changes, and there is no conclusion of 'exponent ', so keep exponent as shell input.
> 
> Should we add both the options? If the user gives time we derive (exponent, mantissa) and if the user gives (exponent, mantissa) we use it directly, WDYT?

My concern is exist multiple sets of exponent and mantissa derived from a given time, and the time calculated from the derived exponent and mantissa may not equal to the given time. Customer may complain about why time set not as expectation.

krish2718 commented:
> > > Hi @krish2718 @jukkar @MaochenWang1
> > > Follow @krish2718 comment to remove teardown changes, and there is no conclusion of 'exponent ', so keep exponent as shell input.
> > 
> > Should we add both the options? If the user gives time we derive (exponent, mantissa) and if the user gives (exponent, mantissa) we use it directly, WDYT?
> 
> My concern is exist multiple sets of exponent and mantissa derived from a given time, and the time calculated from the derived exponent and mantissa may not equal to the given time. Customer may complain about why time set not as expectation.

We share the (exponent, mantissa) in the frame to the AP, so, at least AP and STA will have the same time. Anyways giving both options is better we can leave it to the users to choose the appropriate one, no?

Rex-Chen-NXP commented:
> > > > Hi @krish2718 @jukkar @MaochenWang1
> > > > Follow @krish2718 comment to remove teardown changes, and there is no conclusion of 'exponent ', so keep exponent as shell input.
> > > 
> > > 
> > > Should we add both the options? If the user gives time we derive (exponent, mantissa) and if the user gives (exponent, mantissa) we use it directly, WDYT?
> > 
> > 
> > My concern is exist multiple sets of exponent and mantissa derived from a given time, and the time calculated from the derived exponent and mantissa may not equal to the given time. Customer may complain about why time set not as expectation.
> 
> We share the (exponent, mantissa) in the frame to the AP, so, at least AP and STA will have the same time. Anyways giving both options is better we can leave it to the users to choose the appropriate one, no?

Okay, how can we derive a set of exponent and mantissa from a given time, is there any algorithm for refer?

krish2718 commented:
> > > > > Hi @krish2718 @jukkar @MaochenWang1
> > > > > Follow @krish2718 comment to remove teardown changes, and there is no conclusion of 'exponent ', so keep exponent as shell input.
> > > > 
> > > > 
> > > > Should we add both the options? If the user gives time we derive (exponent, mantissa) and if the user gives (exponent, mantissa) we use it directly, WDYT?
> > > 
> > > 
> > > My concern is exist multiple sets of exponent and mantissa derived from a given time, and the time calculated from the derived exponent and mantissa may not equal to the given time. Customer may complain about why time set not as expectation.
> > 
> > 
> > We share the (exponent, mantissa) in the frame to the AP, so, at least AP and STA will have the same time. Anyways giving both options is better we can leave it to the users to choose the appropriate one, no?
> 
> Okay, how can we derive a set of exponent and mantissa from a given time with high accuracy, is there any algorithm for refer?

We took the simpler approach: https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/wifi/nrf_wifi/src/wifi_mgmt.c#L250

krish2718 commented:
> > > > > > Hi @krish2718 @jukkar @MaochenWang1
> > > > > > Follow @krish2718 comment to remove teardown changes, and there is no conclusion of 'exponent ', so keep exponent as shell input.
> > > > > 
> > > > > 
> > > > > Should we add both the options? If the user gives time we derive (exponent, mantissa) and if the user gives (exponent, mantissa) we use it directly, WDYT?
> > > > 
> > > > 
> > > > My concern is exist multiple sets of exponent and mantissa derived from a given time, and the time calculated from the derived exponent and mantissa may not equal to the given time. Customer may complain about why time set not as expectation.
> > > 
> > > 
> > > We share the (exponent, mantissa) in the frame to the AP, so, at least AP and STA will have the same time. Anyways giving both options is better we can leave it to the users to choose the appropriate one, no?
> > 
> > 
> > Okay, how can we derive a set of exponent and mantissa from a given time with high accuracy, is there any algorithm for refer?
> 
> We took the simpler approach: https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/wifi/nrf_wifi/src/wifi_mgmt.c#L250

Sonnet 3.5 gave this for `binary scaling` input.

```
def convert_time_to_mantissa_exponent(time_ms):
    """
    Convert time in milliseconds to mantissa and exponent representation
    
    Args:
        time_ms (float): Time in milliseconds
    
    Returns:
        tuple: (mantissa, exponent)
    """
    # Start with base-2 exponent of 0
    exponent = 0
    mantissa = time_ms
    
    # Reduce mantissa by dividing by 2 and incrementing exponent
    while mantissa > 255:  # Typical max mantissa value
        mantissa /= 2
        exponent += 1
    
    return (int(mantissa), exponent)

# Example usage
time = 50  # milliseconds
mantissa, exponent = convert_time_to_mantissa_exponent(time)
print(f"Original time: {time}ms")
print(f"Mantissa: {mantissa}")
print(f"Exponent: {exponent}")
print(f"Reconstructed time: {mantissa * (2 ** exponent)}ms")
```

Rex-Chen-NXP commented:
> > > > > > > Hi @krish2718 @jukkar @MaochenWang1
> > > > > > > Follow @krish2718 comment to remove teardown changes, and there is no conclusion of 'exponent ', so keep exponent as shell input.
> > > > > > 
> > > > > > 
> > > > > > Should we add both the options? If the user gives time we derive (exponent, mantissa) and if the user gives (exponent, mantissa) we use it directly, WDYT?
> > > > > 
> > > > > 
> > > > > My concern is exist multiple sets of exponent and mantissa derived from a given time, and the time calculated from the derived exponent and mantissa may not equal to the given time. Customer may complain about why time set not as expectation.
> > > > 
> > > > 
> > > > We share the (exponent, mantissa) in the frame to the AP, so, at least AP and STA will have the same time. Anyways giving both options is better we can leave it to the users to choose the appropriate one, no?
> > > 
> > > 
> > > Okay, how can we derive a set of exponent and mantissa from a given time with high accuracy, is there any algorithm for refer?
> > 
> > 
> > We took the simpler approach: https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/wifi/nrf_wifi/src/wifi_mgmt.c#L250
> 
> Sonnet 3.5 gave this for `binary scaling` input.
> 
> ```
> def convert_time_to_mantissa_exponent(time_ms):
>     """
>     Convert time in milliseconds to mantissa and exponent representation
>     
>     Args:
>         time_ms (float): Time in milliseconds
>     
>     Returns:
>         tuple: (mantissa, exponent)
>     """
>     # Start with base-2 exponent of 0
>     exponent = 0
>     mantissa = time_ms
>     
>     # Reduce mantissa by dividing by 2 and incrementing exponent
>     while mantissa > 255:  # Typical max mantissa value
>         mantissa /= 2
>         exponent += 1
>     
>     return (int(mantissa), exponent)
> 
> # Example usage
> time = 50  # milliseconds
> mantissa, exponent = convert_time_to_mantissa_exponent(time)
> print(f"Original time: {time}ms")
> print(f"Mantissa: {mantissa}")
> print(f"Exponent: {exponent}")
> print(f"Reconstructed time: {mantissa * (2 ** exponent)}ms")
> ```

Thanks for sharing, as Nordic driver implemented the convert of  (exponent, mantissa, interval) , and need interval as argv, to compatible Nordic's exist implementation, the shell should prepare all of parameters(exponent, mantissa, interval) according to the input (exponent, mantissa) or interval, right? If you agree, I will make change.

Rex-Chen-NXP committed:
net: wifi: shell: fix twt setup quick failed issue

The default exponent is 0, will cause twt setup quick failed,
set exponent value derived from twt interval to fix it.

Signed-off-by: Rex Chen <rex.chen_1@nxp.com>

Rex-Chen-NXP commented:
Hi @krish2718 @jukkar  @MaochenWang1 
Follow @krish2718 comments to support both interval  or (exponent, mantissa) as input parameters, format check pass in local, test pass base on RW612 board, could help review after CI check pass?

krish2718 commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -1702,11 +1710,56 @@ static int twt_args_to_params(const struct shell *sh, size_t argc, char *argv[],
+	if ((params->setup.twt_interval != 0) &&
+	   ((params->setup.twt_exponent != 0) ||
+	   (params->setup.twt_mantissa != 0))) {
+		PR_ERROR("Only set interval or exponent and mantissa\n");
```
`Only one of TWT internal or (mantissa, exponent) should be used)`

Rex-Chen-NXP replied:
Yes, the implementation select one of twt interval or (mantissa, exponent), will return failed if input both.

Rex-Chen-NXP replied:
Updated as comment

krish2718 commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -1702,11 +1710,56 @@ static int twt_args_to_params(const struct shell *sh, size_t argc, char *argv[],
+		params->setup.twt_interval = floor(ldexp(params->setup.twt_mantissa,
+							 params->setup.twt_exponent));
+	} else {
+		PR_ERROR("Need set interval or exponent and mantissa\n");
```
`Either TWT interval or (mantissa, exponent) is needed`.

Rex-Chen-NXP replied:
Updated as comment

Rex-Chen-NXP commented:
Hi @krish2718  @jukkar 
As CI check failed early, I rebased for PR, could you help review?

Rex-Chen-NXP commented:
Hi @jukkar @dleach02 
Could you help review and merge PR?
